instrumentation/opentelemetry-instrumentation-sklearn/src/opentelemetry/instrumentation/sklearn/__init__.py

```diff
@@ -0,0 +1,738 @@
+        model = Pipeline(
+            [
+                ("class", RandomForestClassifier(n_estimators=10)),
+            ]
+        )
+
+        model.fit(X_train, y_train)
+
+        SklearnInstrumentor().instrument_estimator(model)
+
+    Args:
+        methods (list): A list of method names on which to instrument a span.
+          This list of methods will be checked on all estimators in the model
+          hierarchy. Used in package and model instrumentation
+        recurse_attribs (dict): A dictionary of ``BaseEstimator``-derived
+          sklearn classes as keys, with values being a list of attributes. Each
+          attribute represents either an estimator or list of estimators on
+          which to also implement spans. An example is
+          ``RandomForestClassifier`` and its attribute ``estimators_``. Used
+          in model instrumentation only.
+        recurse_namedtuple_attribs (dict): A dictionary of ``BaseEstimator``-
+          derived sklearn types as keys, with values being a list of
+          attribute names. Each attribute represents a list of 2-tuples in
+          which the first element is the estimator name, and the second
+          element is the estimator. Defaults include sklearn's ``Pipeline``
+          and its attribute ``steps``, and the ``FeatureUnion`` and its
+          attribute ``transformer_list``. Used in model instrumentation only.
+        spanner: A function with signature (func, str) which
+          decorates instance methods with opentelemetry spans.
+        packages: A list of additional sklearn-compatible packages to
+          instrument. Used with package instrumentation only.
+        exclude_classes: A list of classes to exclude from instrumentation.
+          Child classes are also excluded. Default is sklearn's
+          ``[BaseDecisionTree]``.
+    """
+
+    def __new__(cls, *args, **kwargs):
+        """Override new.
+
+        The base class' new method passes args and kwargs. We override because
+        we init the class with configuration and Python raises TypeError when
+        additional arguments are passed to the object.__new__() method.
+        """
+        if cls._instance is None:
+            cls._instance = object.__new__(cls)
+
+        return cls._instance
+
+    def __init__(
+        self,
+        methods: List[str] = None,
+        recurse_attribs: Dict[Type[BaseEstimator], List[str]] = None,
+        recurse_namedtuple_attribs: Dict[
+            Type[BaseEstimator], List[str]
+        ] = None,
+        spanner: Callable = implement_spans,
+        packages: List[str] = None,
+        exclude_classes: List[Type] = None,
+    ):
+        self.methods = methods or DEFAULT_METHODS
+        self.recurse_attribs = recurse_attribs or DEFAULT_ATTRIBS
+        self.recurse_namedtuple_attribs = (
+            recurse_namedtuple_attribs or DEFAULT_NAMEDTUPLE_ATTRIBS
+        )
+        self.spanner = spanner
+        self.packages = packages or DEFAULT_PACKAGES
```

Comment:
Is this mechanism for if the user wants to instrument other sklearn packages in ADDITION to just `sklearn`? If that's the case, should the two be joined instead?

Reply:
>Is this mechanism for if the user wants to instrument other sklearn packages in ADDITION to just sklearn?

Yes. I left open the use case for when someone might want to instrument an sklearn derivative package, but not sklearn itself. For instance, if the end user wants to instrument _only_ the `lightgbm` package.

Reply:
> packages: A list of additional sklearn-compatible packages to
          instrument. Used with package instrumentation only.

The "additional" part is confusing.

Perhaps the description in the docstring should be updated then?